src/chabochethermal.jl

```diff
@@ -0,0 +1,1050 @@
+    #   ΔR = b (Q - R_new) Δp               (component 7)
+    #   ΔXj = ((2/3) Cj n - Dj Xj_new) Δp   (components 8:13, 14:19, 20:25) (no sum)
+    #
+    # where
+    #
+    #   Δ(...) = (...)_new - (...)_old
+    #
+    # (Δp and n are described below.)
+    #
+    # Then in each equation, move the terms on the RHS to the LHS to get
+    # the standard form, (stuff) = 0. Then the LHS is the residual.
+    #
+    # The viscoplastic response is updated by:
+    #
+    #   Δεpl = Δp n
+    #
+    # where
+    #
+    #   Δp = p' Δt
+    #   p' = 1/tvp * (<f> / Kn)^nn  (Norton-Bailey power law; <...>: Macaulay brackets)
+    #   f = √(3/2 dev(σ_eff) : dev(σ_eff)) - (R0 + R)
+    #   σ_eff = σ - ∑ Xj
+    #   n = ∂f/∂σ
+    #
+    # `F` is output, length 25.
+    function r!(F::V, stress_new::Symm2{<:Real}, R_new::Real,
+                X1_new::Symm2{<:Real}, X2_new::Symm2{<:Real}, X3_new::Symm2{<:Real},
+                dstrain::Symm2{<:Real}, temperature_new::Real) where {V <: AbstractVector{<:Real}}
+        # This stuff must be done here so we can autodiff it w.r.t. temperature_new.
+        thermal_strainf(theta) = thermal_strain_tensor(alphaf, theta0, theta)
+        # thermal_strain_derivative(theta) = gradient(thermal_strainf, theta)
+        # thermal_dstrain = thermal_strain_derivative(temperature_new) * (temperature_new - temperature)
+        thermal_dstrain = thermal_strainf(temperature_new) - thermal_strainf(temperature)
+
+        Df(theta) = elasticity_tensor(Ef, nuf, theta)  # dσ/dε, i.e. ∂σij/∂εkl
+        dDdthetaf(theta) = gradient(Df, theta)
+        D = Df(temperature_new)
+        dDdtheta = dDdthetaf(temperature_new)
+
+        dotp = dotpf(stress_new, R_new, X1_new, X2_new, X3_new, temperature_new)
+        n = nf(stress_new, R_new, X1_new, X2_new, X3_new, temperature_new)
+
+        local dtemperature = temperature_new - temperature
+
+        dp = dotp * dtime
+        plastic_dstrain = dp * n
+        elastic_dstrain = dstrain - plastic_dstrain - thermal_dstrain
+        elastic_strain = elastic_strain_old + elastic_dstrain
+        tovoigt!(view(F, 1:6),
+                 stress_new - stress
+                 - dcontract(D, elastic_dstrain)
+                 - dcontract(dDdtheta, elastic_strain) * dtemperature)
+
+        # Reijo's equations (37) and (43), for exponentially saturating
+        # isotropic hardening, are:
+        #
+        #   Kiso = Kiso∞ (1 - exp(-hiso κiso / Kiso∞))
+        #   κiso' = 1 / tvp <fhat / σ0>^p
+        #
+        # Our equation for R in the case without thermal effects, where
+        # Q and b are constant, is:
+        #
+        #   R' = b (Q - R) p'
+        #
+        # We identify (LHS Reijo's notation; RHS Materials.jl notation):
+        #
+        #   Kiso = R, κiso = p, σ0 = Kn, p = nn
+        #   TODO: is fhat our f? Looks a bit different.
+        #
+        # So in the notation used in Materials.jl:
+        #
+        #   R = R∞ (1 - exp(-hiso p / R∞))
```

Comment:
Sometimes the R_0 is included in the integrated version so that R describes the total size of the elastic region (in Chaboche's notation):
<img src="https://render.githubusercontent.com/render/math?math=%5Cbegin%7Bequation%7D%0AR%20%3D%20R_0%20%2B%20%5Csum_%7Bi%3D1%7D%5EN%20Q_i%20%5Cleft(%201-%5Cexp%5Cleft%5B-b_i%20p%20%5Cright%5D%5Cright)%0A%5Cend%7Bequation%7D">
Meaning Q_i is the saturation value of isotropic hardening (in units of stresses) and b_i is the corresponding saturation coefficient of each component. Then in the more general version we have strain-amplitude (modelled as 'memory surface' radius q) dependent hardening saturation values Q_i(q).

Reply:
Ah, thanks.

A clarification: the comments in `create_nonlinear_system_of_equations` that talk about *"Reijo's equations"*, refer to the team-internal memo `materials_modelling.pdf` that Reijo wrote and sent us last year, when we were beginning the current stage of development.

I think that at some point, some of that content should go into a user manual for `Materials.jl` (in a spirit similar to how my [user manual for pydgq](https://github.com/Technologicat/pydgq/blob/master/doc/pydgq_user_manual.pdf) explains time-discontinuous Galerkin).